### PR TITLE
feat(turbopack): Sort issues

### PR DESCRIPTION
**Pull request by kdy1**
### What?

Sort issues emitted by turbopack.

### Why?

Make it deterministic and testable

### How?

Closes PACK-2440

turbopack counterpart: https://github.com/vercel/turbo/pull/7519

### CODE REVIEW

**Comment by ijjk** [comment]
# Stats from current PR

<details open>
<summary><strong>Default Build</strong></summary>

<br/>

<details>
<summary><strong>General</strong></summary>

|  | vercel/next.js canary  | vercel/next.js kdy1/pack-2440 | Change |
| - | - | - | - |
| buildDuration | 13.8s | 13.9s | N/A |
| buildDurationCached | 7.6s | 6.5s | N/A |
| nodeModulesSize | 197 MB | 197 MB | N/A |
| nextStartRea..uration (ms) | 406ms | 406ms | ✓ |

</details>

<details>
<summary><strong>Client Bundles (main, webpack)</strong></summary>

|  | vercel/next.js canary  | vercel/next.js kdy1/pack-2440 | Change |
| - | - | - | - |
| 2453-HASH.js gzip | 30.5 kB | 30.5 kB | N/A |
| 3304.HASH.js gzip | 181 B | 181 B | ✓ |
| 3f784ff6-HASH.js gzip | 53.7 kB | 53.7 kB | N/A |
| 8299-HASH.js gzip | 5.04 kB | 5.04 kB | N/A |
| framework-HASH.js gzip | 45.2 kB | 45.2 kB | ✓ |
| main-app-HASH.js gzip | 242 B | 242 B | ✓ |
| main-HASH.js gzip | 32.1 kB | 32.2 kB | N/A |
| webpack-HASH.js gzip | 1.7 kB | 1.7 kB | N/A |
| Overall change | 45.6 kB | 45.6 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | vercel/next.js kdy1/pack-2440 | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | vercel/next.js kdy1/pack-2440 | Change |
| - | - | - | - |
| _app-HASH.js gzip | 196 B | 197 B | N/A |
| _error-HASH.js gzip | 184 B | 184 B | ✓ |
| amp-HASH.js gzip | 505 B | 505 B | ✓ |
| css-HASH.js gzip | 324 B | 325 B | N/A |
| dynamic-HASH.js gzip | 2.5 kB | 2.5 kB | N/A |
| edge-ssr-HASH.js gzip | 258 B | 258 B | ✓ |
| head-HASH.js gzip | 352 B | 352 B | ✓ |
| hooks-HASH.js gzip | 370 B | 371 B | N/A |
| image-HASH.js gzip | 4.2 kB | 4.2 kB | ✓ |
| index-HASH.js gzip | 259 B | 259 B | ✓ |
| link-HASH.js gzip | 2.67 kB | 2.67 kB | N/A |
| routerDirect..HASH.js gzip | 314 B | 312 B | N/A |
| script-HASH.js gzip | 386 B | 386 B | ✓ |
| withRouter-HASH.js gzip | 309 B | 309 B | ✓ |
| 1afbb74e6ecf..834.css gzip | 106 B | 106 B | ✓ |
| Overall change | 6.56 kB | 6.56 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | vercel/next.js kdy1/pack-2440 | Change |
| - | - | - | - |
| _buildManifest.js gzip | 483 B | 485 B | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | vercel/next.js kdy1/pack-2440 | Change |
| - | - | - | - |
| index.html gzip | 527 B | 527 B | ✓ |
| link.html gzip | 540 B | 540 B | ✓ |
| withRouter.html gzip | 523 B | 522 B | N/A |
| Overall change | 1.07 kB | 1.07 kB | ✓ |

</details>

<details>
<summary><strong>Edge SSR bundle Size</strong></summary>

|  | vercel/next.js canary  | vercel/next.js kdy1/pack-2440 | Change |
| - | - | - | - |
| edge-ssr.js gzip | 95 kB | 95 kB | N/A |
| page.js gzip | 3.06 kB | 3.06 kB | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Middleware size</strong></summary>

|  | vercel/next.js canary  | vercel/next.js kdy1/pack-2440 | Change |
| - | - | - | - |
| middleware-b..fest.js gzip | 625 B | 622 B | N/A |
| middleware-r..fest.js gzip | 151 B | 151 B | ✓ |
| middleware.js gzip | 25.5 kB | 25.5 kB | N/A |
| edge-runtime..pack.js gzip | 839 B | 839 B | ✓ |
| Overall change | 990 B | 990 B | ✓ |

</details>

<details>
<summary><strong>Next Runtimes</strong></summary>

|  | vercel/next.js canary  | vercel/next.js kdy1/pack-2440 | Change |
| - | - | - | - |
| app-page-exp...dev.js gzip | 171 kB | 171 kB | ✓ |
| app-page-exp..prod.js gzip | 96.7 kB | 96.7 kB | ✓ |
| app-page-tur..prod.js gzip | 98.5 kB | 98.5 kB | ✓ |
| app-page-tur..prod.js gzip | 92.9 kB | 92.9 kB | ✓ |
| app-page.run...dev.js gzip | 150 kB | 150 kB | ✓ |
| app-page.run..prod.js gzip | 91.4 kB | 91.4 kB | ✓ |
| app-route-ex...dev.js gzip | 21.3 kB | 21.3 kB | ✓ |
| app-route-ex..prod.js gzip | 15 kB | 15 kB | ✓ |
| app-route-tu..prod.js gzip | 15 kB | 15 kB | ✓ |
| app-route-tu..prod.js gzip | 14.8 kB | 14.8 kB | ✓ |
| app-route.ru...dev.js gzip | 20.9 kB | 20.9 kB | ✓ |
| app-route.ru..prod.js gzip | 14.8 kB | 14.8 kB | ✓ |
| pages-api-tu..prod.js gzip | 9.51 kB | 9.51 kB | ✓ |
| pages-api.ru...dev.js gzip | 9.79 kB | 9.79 kB | ✓ |
| pages-api.ru..prod.js gzip | 9.51 kB | 9.51 kB | ✓ |
| pages-turbo...prod.js gzip | 22.3 kB | 22.3 kB | ✓ |
| pages.runtim...dev.js gzip | 23 kB | 23 kB | ✓ |
| pages.runtim..prod.js gzip | 22.3 kB | 22.3 kB | ✓ |
| server.runti..prod.js gzip | 50.6 kB | 50.6 kB | ✓ |
| Overall change | 950 kB | 950 kB | ✓ |

</details>

<details>
<summary><strong>build cache</strong></summary>

|  | vercel/next.js canary  | vercel/next.js kdy1/pack-2440 | Change |
| - | - | - | - |
| 0.pack gzip | 1.56 MB | 1.56 MB | N/A |
| index.pack gzip | 105 kB | 104 kB | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Diff details</strong></summary>

<details>
<summary>Diff for <strong>middleware.js</strong></summary>

Diff too large to display
</details>

</details>

</details>
Commit: 2132e9aa3c0a9835c6e9f18a56e61b0f1882ec7a

File: packages/next-swc/crates/next-core/src/next_font/issue.rs [hunk 1]
```diff
@@ -19,6 +19,11 @@ impl Issue for NextFontIssue {
         Vc::cell("other".to_string())
     }
 
+    #[turbo_tasks::function]
+    fn stage(&self) -> Vc<IssueStage> {
+        IssueStage::Misc.into()
```
Comment (sokra) [inline packages/next-swc/crates/next-core/src/next_font/issue.rs:24]:
```suggestion
        IssueStage::CodeGen.into()
```

File: packages/next-swc/crates/next-core/src/next_server/resolve.rs [hunk 2]
```diff
@@ -436,6 +436,11 @@ impl Issue for UnableToExternalize {
         Vc::cell("externals".to_string())
     }
 
+    #[turbo_tasks::function]
+    fn stage(&self) -> Vc<IssueStage> {
+        IssueStage::Load.into()
```
Comment (sokra) [inline packages/next-swc/crates/next-core/src/next_server/resolve.rs:441]:
```suggestion
        IssueStage::Config.into()
```

File: packages/next-swc/crates/next-core/src/next_shared/resolve.rs [hunk 3]
```diff
@@ -126,6 +126,11 @@ impl Issue for InvalidImportModuleIssue {
         Vc::cell("resolve".to_string())
     }
 
+    #[turbo_tasks::function]
+    fn stage(&self) -> Vc<IssueStage> {
+        IssueStage::Analysis.into()
```
Comment (sokra) [inline packages/next-swc/crates/next-core/src/next_shared/resolve.rs:131]:
```suggestion
        IssueStage::Resolve.into()
```

**Comment by ijjk** [comment]
## Failing test suites 
Commit: 35b41ef7fa21822f15cbc402765c902686049005

`pnpm test-start test/e2e/app-dir/navigation/navigation.test.ts`
- app dir - navigation > navigation between pages and app > should not continously initiate a mpa navigation to the same URL when router state changes

<details>
<summary>Expand output</summary>

  ● app dir - navigation › navigation between pages and app › should not continously initiate a mpa navigation to the same URL when router state changes

    page.waitForSelector: Timeout 60000ms exceeded.
    Call log:
      - waiting for locator('#link-to-slow-page')
      -   locator resolved to visible <a href="/slow-page" id="link-to-slow-page">To /slow-page</a>

      421 |     return this.chain(() => {
      422 |       return page
    > 423 |         .waitForSelector(selector, { timeout, state: 'attached' })
          |          ^
      424 |         .then(async (el) => {
      425 |           // it seems selenium waits longer and tests rely on this behavior
      426 |           // so we wait for the load event fire before returning

      at waitForSelector (lib/browsers/playwright.ts:423:10)

</details>


Read more about building and testing Next.js in [contributing.md](https://github.com/vercel/next.js/blob/canary/contributing.md).

**Review by sokra — APPROVED** [review]

**Comment by sokra** [comment]
> I think the test failure here is real and caused by the issue sorting

That test case was flaky before... So it being consistent is an improvement at least, even while the error message is less useful^^

**Comment by ForsakenHarmony** [comment]
@sokra it's the wrong error message though afaik, I'd rather disable the test 🤔

**Comment by ForsakenHarmony** [comment]
at the very least, leave the correct snapshot there commented out and add a todo comment

**Review by ForsakenHarmony — CHANGES_REQUESTED** [review]
see above

this whole problem is caused by the insistence to continue compiling after errors have occurred

**Review by ForsakenHarmony — APPROVED** [review]

**Comment by sokra** [comment]
> timestamps

timestamps won't work when everything is parallel and might recompute independently.

Anyway, so let's change the ordering so that Resolving is after Analysis

**Pull request merged**